File: nodes/halo_radar_node.py
#!/usr/bin/env python

import rospy
from marine_msgs.msg import RadarSectorStamped
from marine_msgs.msg import RadarScanline
from marine_msgs.msg import RadarControlItem
from marine_msgs.msg import RadarControlSet
from marine_msgs.msg import KeyValue
import halo_radar.halo_radar
import datetime

class RadarPublisher:
    def __init__(self, prefix, radar):
        self.data_pub = rospy.Publisher(prefix+'/data',RadarSectorStamped,queue_size=10)
        self.state_pub = rospy.Publisher(prefix+'/state',RadarControlSet,queue_size=10)
        self.radar = radar
        self.radar.state_callback = self.onState
        self.radar.state_update_period = 2
        self.radar.data_callback = self.onData
        
        self.state_change_subscriber = rospy.Subscriber(prefix+'/change_state', KeyValue, self.state_change_callback)

    def createEnumControl(self, name, value, label, enums):
        rci = RadarControlItem()
        rci.name = name
        rci.value = value
        rci.label = label
        rci.type = RadarControlItem.CONTROL_TYPE_ENUM
        for e in enums:
            rci.enums.append(e)
        return rci
    
    def createFloatControl(self, name, value, label, min_value = 0.0, max_value = 0.0):
        rci = RadarControlItem()
        rci.name = name
        rci.value = str(value)
        rci.label = label
        rci.type = RadarControlItem.CONTROL_TYPE_FLOAT
        rci.min_value = min_value
        rci.max_value = max_value
        return rci
    
    def createFloatWithAutoControl(self, name, value, label, min_value = 0.0, max_value = 0.0):
        rci = RadarControlItem()
        rci.name = name
        rci.value = value
        rci.label = label
        rci.type = RadarControlItem.CONTROL_TYPE_FLOAT_WITH_AUTO
        rci.min_value = min_value
        rci.max_value = max_value
        return rci

    def onState(self,state):
        rcs = RadarControlSet()
        if 'status' in state:
            rcs.items.append(self.createEnumControl('status',state['status'],'Status',('standby','transmit')))
        if 'range' in state:
            rcs.items.append(self.createFloatControl('range',state['range'],'Range',25,75000))
        if 'mode' in state:
            rcs.items.append(self.createEnumControl('mode',state['mode'],'Mode',('custom','harbor','offshore','weather','bird')))
        if 'gain' in state and 'gain_mode' in state:
            if state['gain_mode'] == 'auto':
                value = 'auto'
            else:
                value = str(state['gain'])
            rcs.items.append(self.createFloatWithAutoControl('gain',value,'Gain',0,100))
        if 'sea_clutter' in state and 'sea_clutter_mode' in state:
            if state['sea_clutter_mode'] == 'auto':
                value = 'auto'
            else:
                value = str(state['sea_clutter'])
            rcs.items.append(self.createFloatWithAutoControl('sea_clutter',value,'Sea clutter',0,100))
        if 'auto_sea_clutter_nudge' in state:
            rcs.items.append(self.createFloatControl('auto_sea_clutter_nudge',state['auto_sea_clutter_nudge'],'Auto sea clutter nudge',-50,50))
        if 'sea_state' in state:
            rcs.items.append(self.createEnumControl('sea_state',state['sea_state'],'Sea state',('calm','moderate','rough')))
        if 'rain_clutter' in state:
            rcs.items.append(self.createFloatControl('rain_clutter',state['rain_clutter'],'Rain clutter',0,100))
        if 'noise_rejection' in state:
            rcs.items.append(self.createEnumControl('noise_rejection',state['noise_rejection'],'Noise rejection',('off','low','medium','high')))
        if 'target_expansion' in state:
            rcs.items.append(self.createEnumControl('target_expansion',state['target_expansion'],'Target expansion',('off','low','medium','high')))
        if 'interference_rejection' in state:
            rcs.items.append(self.createEnumControl('interference_rejection',state['interference_rejection'],'Interf. rej',('off','low','medium','high')))
        if 'target_separation' in state:
            rcs.items.append(self.createEnumControl('target_separation',state['target_separation'],'Target separation',('off','low','medium','high')))
        if 'scan_speed' in state:
            rcs.items.append(self.createEnumControl('scan_speed',state['scan_speed'],'Fast scan',('off','medium','high')))
        if 'doppler_mode' in state:
            rcs.items.append(self.createEnumControl('doppler_mode',state['doppler_mode'],'VelocityTrack',('off','normal','approaching_only')))
        if 'doppler_speed' in state:
            rcs.items.append(self.createFloatControl('doppler_speed',state['doppler_speed'],'Speed Threshold',0.05,15.95))
        if 'antenna_height' in state:
            rcs.items.append(self.createFloatControl('antenna_height',state['antenna_height'],'Antenna height',0.0,30.175))
        if 'bearing_alignment' in state:
            rcs.items.append(self.createFloatControl('bearing_alignment',state['bearing_alignment'],'Bearing alignment',0,365))
        if 'sidelobe_suppression' in state and 'sidelobe_suppression_mode' in state:
            if state['sidelobe_suppression_mode'] == 'auto':
                value = 'auto'
            else:
                value = str(state['sidelobe_suppression'])
            rcs.items.append(self.createFloatWithAutoControl('sidelobe_suppression',value,'Sidelobe sup.',0,100))
        if 'light' in state:
            rcs.items.append(self.createEnumControl('light',state['light'],'Halo light',('off','low','medium','high')))

        self.state_pub.publish(rcs)

    def onData(self,data):
        if data is not None:
            sector = RadarSectorStamped()
            sector.header.stamp = rospy.get_rostime()
            sector.header.frame_id = 'radar'

            for sl in data['scanlines']:
                scanline = RadarScanline()
                scanline.angle = sl['angle']
                scanline.range = sl['range']
                scanline.intensities = sl['intensities']
                sector.sector.scanlines.append(scanline)
            self.data_pub.publish(sector)
            
    def state_change_callback(self,data):
        self.radar.sendCommand(data.key,data.value)

if __name__ == '__main__':
    rospy.init_node('halo_radar')
    found_radars = halo_radar.halo_radar.scan()
    radars = []
    if found_radars is not None:
        i = found_radars[0]
        print i
        for r in found_radars[1]:
            print '\t',r
            radars.append(RadarPublisher('/base/radar/'+r['label'], halo_radar.halo_radar.HaloRadar(i,r['data'],r['send'],r['report'],r['label'])))

        #for r in radars:
            #r.radar.on()
        while not rospy.is_shutdown():
            for r in radars:
                r.radar.checkSockets()

File: src/halo_radar/halo_radar_node.cpp
#include <ros/ros.h>
#include <tf2/utils.h>
#include <iostream>
#include "halo_radar.h"
#include "marine_sensor_msgs/RadarSector.h"
#include "marine_radar_control_msgs/RadarControlSet.h"
#include "marine_radar_control_msgs/RadarControlValue.h"
#include "nav_msgs/Odometry.h"
#include <future>
#include "angular_speed_estimator.h"

class RosRadar : public halo_radar::Radar
{
public:
  RosRadar(halo_radar::AddressSet const &addresses) : halo_radar::Radar(addresses)
  {
    ros::NodeHandle n;
    m_data_pub = n.advertise<marine_sensor_msgs::RadarSector>(addresses.label + "/data", 10);
    m_state_pub = n.advertise<marine_radar_control_msgs::RadarControlSet>(addresses.label + "/state", 10);
    m_state_change_sub =
        n.subscribe(addresses.label + "/change_state", 10, &RosRadar::stateChangeCallback, this);
    m_heartbeatTimer = n.createTimer(ros::Duration(1.0), &RosRadar::hbTimerCallback, this);
    ros::param::param<double>("~range_correction_factor", m_rangeCorrectionFactor, m_rangeCorrectionFactor);
    ros::param::param<std::string>("~frameId", m_frame_id, m_frame_id);

    startThreads();
  }

 protected:
  void processData(std::vector<halo_radar::Scanline> const &scanlines) override
  {
    if(scanlines.empty())
      return;
    marine_sensor_msgs::RadarSector rs;
    rs.header.stamp = ros::Time::now();
    rs.header.frame_id = m_frame_id;
    rs.angle_start = 2.0*M_PI*(360-scanlines.front().angle)/360.0;
    double  angle_max = 2.0*M_PI*(360-scanlines.back().angle)/360.0;
    if(scanlines.size() > 1)
    {
      if (angle_max > rs.angle_start && angle_max-rs.angle_start > M_PI) // have we looped around (also make sure angle are decreasing)
        angle_max -= 2.0*M_PI;
      rs.angle_increment = (angle_max-rs.angle_start)/double(scanlines.size()-1);

    }
    rs.range_min = 0.0;
    rs.range_max = scanlines.front().range;
    for (auto sl : scanlines)
    {
      marine_sensor_msgs::RadarEcho echo;
      for (auto i : sl.intensities)
        echo.echoes.push_back(i/15.0); // 4 bit int to float
      rs.intensities.push_back(echo);
    }

    auto angular_speed = m_estimator.update(rs.header.stamp,  rs.angle_start);
    double scan_time = 0.0;
    if(angular_speed != 0.0)
      scan_time = 2*M_PI/fabs(angular_speed);

    rs.scan_time = ros::Duration(scan_time);

    double time_increment = 0.0;
    if (scan_time > 0)
      time_increment = std::abs(rs.angle_increment)/scan_time;
    rs.time_increment =  ros::Duration(time_increment);

    m_data_pub.publish(rs);
  }

  void stateUpdated() override
  {
    marine_radar_control_msgs::RadarControlSet rcs;

    std::string statusEnums[] = {"standby", "transmit", ""};

    createEnumControl("status", "Status", statusEnums, rcs);
    createFloatControl("range", "Range", 25, 75000, rcs);

    std::string modeEnums[] = {"custom", "harbor", "offshore", "weather", "bird", ""};

    createEnumControl("mode", "Mode", modeEnums, rcs);
    createFloatWithAutoControl("gain", "gain_mode", "Gain", 0, 100, rcs);
    createFloatWithAutoControl("sea_clutter", "sea_clutter_mode", "Sea clutter", 0, 100, rcs);
    createFloatControl("auto_sea_clutter_nudge", "Auto sea clut adj", -50, 50, rcs);

    std::string seaStateEnums[] = {"calm", "moderate", "rough", ""};

    createEnumControl("sea_state", "Sea state", seaStateEnums, rcs);
    createFloatControl("rain_clutter", "Rain clutter", 0, 100, rcs);

    std::string lowMedHighEnums[] = {"off", "low", "medium", "high", ""};

    createEnumControl("noise_rejection", "Noise rejection", lowMedHighEnums, rcs);
    createEnumControl("target_expansion", "Target expansion", lowMedHighEnums, rcs);
    createEnumControl("interference_rejection", "Interf. rej", lowMedHighEnums, rcs);
    createEnumControl("target_separation", "Target separation", lowMedHighEnums, rcs);

    std::string scanSpeedEnums[] = {"off", "medium", "high", ""};

    createEnumControl("scan_speed", "Fast scan", scanSpeedEnums, rcs);

    std::string dopplerModeEnums[] = {"off", "normal", "approaching_only", ""};

    createEnumControl("doppler_mode", "VelocityTrack", dopplerModeEnums, rcs);
    createFloatControl("doppler_speed", "Speed threshold", 0.05, 15.95, rcs);
    createFloatControl("antenna_height", "Antenna height", 0.0, 30.175, rcs);
    createFloatControl("bearing_alignment", "Bearing alignment", 0, 360, rcs);
    createFloatWithAutoControl("sidelobe_suppression", "sidelobe_suppression_mode", "Sidelobe sup.", 0, 100, rcs);
    createEnumControl("lights", "Halo light", lowMedHighEnums, rcs);

    m_state_pub.publish(rcs);
  }

 private:
  void stateChangeCallback(const marine_radar_control_msgs::RadarControlValue::ConstPtr &cv)
  {
    sendCommand(cv->key, cv->value);
  }

  void hbTimerCallback(const ros::TimerEvent &e)
  {
    if (checkHeartbeat())
      stateUpdated();
  }

  void createEnumControl(std::string const &name, std::string const &label, std::string const enums[],
                         marine_radar_control_msgs::RadarControlSet &rcs)
  {
    if (m_state.find(name) != m_state.end())
    {
      marine_radar_control_msgs::RadarControlItem rci;
      rci.name = name;
      rci.value = m_state[name];
      rci.label = label;
      rci.type = marine_radar_control_msgs::RadarControlItem::CONTROL_TYPE_ENUM;
      for (int i = 0; !enums[i].empty(); i++)
        rci.enums.push_back(enums[i]);
      rcs.items.push_back(rci);
    }
  }

  void createFloatControl(std::string const &name, std::string const &label, float min_value, float max_value,
                          marine_radar_control_msgs::RadarControlSet &rcs)
  {
    if (m_state.find(name) != m_state.end())
    {
      marine_radar_control_msgs::RadarControlItem rci;
      rci.name = name;
      rci.value = m_state[name];
      rci.label = label;
      rci.type = marine_radar_control_msgs::RadarControlItem::CONTROL_TYPE_FLOAT;
      rci.min_value = min_value;
      rci.max_value = max_value;
      rcs.items.push_back(rci);
    }
  }

  void createFloatWithAutoControl(std::string const &name, std::string const &auto_name, std::string const &label,
                                  float min_value, float max_value, marine_radar_control_msgs::RadarControlSet &rcs)
  {
    if (m_state.find(name) != m_state.end() && m_state.find(auto_name) != m_state.end())
    {
      marine_radar_control_msgs::RadarControlItem rci;
      rci.name = name;
      std::string value = m_state[name];
      if (m_state[auto_name] == "auto")
        value = "auto";
      rci.value = value;
      rci.label = label;
      rci.type = marine_radar_control_msgs::RadarControlItem::CONTROL_TYPE_FLOAT_WITH_AUTO;
      rci.min_value = min_value;
      rci.max_value = max_value;
      rcs.items.push_back(rci);
    }
  }

  ros::Publisher m_data_pub;
  ros::Publisher m_state_pub;
  ros::Subscriber m_state_change_sub;
  ros::Timer m_heartbeatTimer;

  double m_rangeCorrectionFactor = 1.024;
  std::string m_frame_id = "radar";

  AngularSpeedEstimator m_estimator;
};

std::shared_ptr<halo_radar::HeadingSender> headingSender;

void odometryCallback(const nav_msgs::Odometry::ConstPtr msg)
{
  if(headingSender)
  {
    double heading = 90.0-180.0*tf2::getYaw(msg->pose.pose.orientation)/M_PI;
    headingSender->setHeading(heading);
  }
    
}

int main(int argc, char **argv)
{
  ros::init(argc, argv, "halo_radar");
  std::vector<std::shared_ptr<RosRadar> > radars;
  std::vector<uint32_t> hostIPs;
  if (ros::param::has("~hostIPs"))
  {
    std::vector<std::string> hostIPstrings;
    ros::param::get("~hostIPs", hostIPstrings);
    for (auto s: hostIPstrings)
      hostIPs.push_back(halo_radar::ipAddressFromString(s));
  }

  std::future<void> scanResult = std::async(std::launch::async, [&] {
    while(radars.empty())
    {
      std::vector<halo_radar::AddressSet> as;
      if(hostIPs.empty())
        as = halo_radar::scan();
      else
        as = halo_radar::scan(hostIPs);
      if(as.empty())
        ROS_WARN_STREAM("No radars found!");
      for (auto a : as)
      {
        radars.push_back(std::shared_ptr<RosRadar>(new RosRadar(a)));
        if(!headingSender)
          headingSender = std::shared_ptr<halo_radar::HeadingSender>(new halo_radar::HeadingSender(a.interface));
      }
    }
  });

  ros::spin();

  return 0;
}
